### PR TITLE
Single feedback topic

### PR DESCRIPTION
**Pull request by laurence-bird**

### CODE REVIEW

File: src/main/scala/com/ovoenergy/orchestration/domain/BuildFeedback.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    }
+  }
+
+  def extractCustomer(deliverTo: DeliverTo): Option[Customer] = {
+    deliverTo match {
+      case customer: Customer => Some(customer)
+      case _                  => None
+    }
+  }
+
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/domain/BuildFeedback.scala]: Type signatures on these implicits please (a bit tedious, sorry)

File: src/main/scala/com/ovoenergy/orchestration/kafka/IssueFeedback.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package com.ovoenergy.orchestration.kafka
+
+import cats.effect.Async
+import cats.syntax.flatMap._
+import com.ovoenergy.comms.model.{FailedV3, Feedback, InternalMetadata, MetadataV3}
+import com.ovoenergy.comms.templates.util.Hash
+import com.ovoenergy.orchestration.domain.{BuildFeedback, FailureDetails}
+import org.apache.kafka.clients.producer.RecordMetadata
+
+class IssueFeedback[F[_]: Async](sendFeedback: Feedback => F[RecordMetadata],
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/IssueFeedback.scala]: There's no reason to put `Async` here, and it limits the possible implementations you can have

File: src/main/scala/com/ovoenergy/orchestration/kafka/IssueFeedback.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+                                 sendFailed: FailedV3 => F[RecordMetadata]) {
+  def send[T](t: T)(implicit buildFeedback: BuildFeedback[T]): F[RecordMetadata] = {
+    val feedback: Feedback = buildFeedback(t)
+    sendFeedback(feedback)
+  }
+
+  // TODO this should go away once we stop producing to old feedback topic
+  def sendWithLegacy(failureDetails: FailureDetails,
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/IssueFeedback.scala]: you can just add a `FlatMap` or `Apply` constraint here (depending on whether you want to use `>>` or `*>`

File: src/main/scala/com/ovoenergy/orchestration/Main.scala [hunk 4]
```diff
@@ -199,14 +192,11 @@ object Main extends StreamApp[IO] with LoggingWithMDC with ExecutionContexts {
       sendFailedCancellationEvent = sendFailedCancellationEvent,
       sendSuccessfulCancellationEvent = sendCancelledEvent,
       descheduleComm = descheduleComm,
-      generateTraceToken = () => UUID.randomUUID().toString
+      generateTraceToken = () => UUID.randomUUID().toString,
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/Main.scala]: Not a point for this PR, but these should probably all be `IO`/`F` instead across the project

File: src/main/scala/com/ovoenergy/orchestration/domain/package.scala [hunk 5]
```diff
@@ -77,4 +77,41 @@ package object domain {
...
+  case class FailureDetails(deliverTo: DeliverTo,
+                            commId: CommId,
+                            traceToken: TraceToken,
+                            eventId: EventId,
+                            reason: String,
+                            errorCode: ErrorCode,
+                            failureType: FailureType)
+
+  object FailureDetails {
+    def apply(deliverTo: DeliverTo,
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/domain/package.scala]: Having an `apply` that takes String _somewhat_ weakens the impact of having typed `CommId` `TraceToken` and `EventId`

File: src/main/scala/com/ovoenergy/orchestration/kafka/Producer.scala [hunk 6]
```diff
@@ -83,4 +86,17 @@ object Producer {
...
   }
+
+  def publisherFor[E: Loggable](topic: Topic[E], key: E => String)(implicit schemaFor: SchemaFor[E],
+                                                                   toRecord: ToRecord[E],
+                                                                   classTag: ClassTag[E]): E => IO[RecordMetadata] = {
+    val producer: KafkaProducer[String, E] = exitAppOnFailure(Producer[E](topic), topic.name)
+    val publisher = { e: E =>
+      Producer
+        .publisher[E](key, producer, topic.name)(e)
+        .flatMap((rm: RecordMetadata) => IO(info((rm, e))(s"Sent event to ${topic.name}")) >> IO.pure(rm))
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/Producer.scala]: This pattern (operating on an element with an `E => F[Unit]`, then returning the original element) is expressed by a very convenient function called `flatTap`, so:

```scala
publisher(...)
  .flatTap(rm => IO(info(rm))
```

File: src/main/scala/com/ovoenergy/orchestration/kafka/consumers/CancellationRequestConsumer.scala [hunk 7]
```diff
@@ -1,20 +1,26 @@
...
 import com.ovoenergy.orchestration.logging.LoggingWithMDC
 import com.ovoenergy.orchestration.processes.Orchestrator.ErrorDetails
 import org.apache.kafka.clients.producer.RecordMetadata
 
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.ExecutionContext
 
 object CancellationRequestConsumer extends LoggingWithMDC {
 
   def apply[F[_]: Async](
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/consumers/CancellationRequestConsumer.scala]: That `Async` constraint is too powerful, as far as I can see from here.

File: src/main/scala/com/ovoenergy/orchestration/kafka/consumers/CancellationRequestConsumer.scala [hunk 8]
```diff
@@ -1,20 +1,26 @@
...
-      sendFailedCancellationEvent: (FailedCancellationV3) => F[RecordMetadata],
-      sendSuccessfulCancellationEvent: (CancelledV3 => F[RecordMetadata]),
+      sendFailedCancellationEvent: FailedCancellationV3 => F[RecordMetadata],
+      sendSuccessfulCancellationEvent: CancelledV3 => F[RecordMetadata],
       descheduleComm: CancellationRequestedV3 => Seq[Either[ErrorDetails, MetadataV3]],
-      generateTraceToken: () => String)(implicit ec: ExecutionContext): CancellationRequestedV3 => F[Unit] = {
+      generateTraceToken: () => String,
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/consumers/CancellationRequestConsumer.scala]: Not for this PR probably, but this should be `F[TraceToken]` instead

File: src/main/scala/com/ovoenergy/orchestration/kafka/consumers/CancellationRequestConsumer.scala [hunk 9]
```diff
@@ -24,15 +30,29 @@ object CancellationRequestConsumer extends LoggingWithMDC {
         val result: Seq[F[RecordMetadata]] = descheduleComm(cancellationRequest).map {
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/consumers/CancellationRequestConsumer.scala]: We need to get into the habit of not using `Seq`

File: src/main/scala/com/ovoenergy/orchestration/kafka/consumers/TriggeredConsumer.scala [hunk 10]
```diff
@@ -1,69 +1,100 @@
...
+import com.ovoenergy.orchestration.domain.{FailureDetails, InternalFailure}
+import com.ovoenergy.orchestration.kafka.IssueFeedback
 import org.apache.kafka.clients.producer.RecordMetadata
-import scala.concurrent.{ExecutionContext}
+
+import scala.concurrent.ExecutionContext
 
 object TriggeredConsumer extends LoggingWithMDC {
 
   def apply[F[_]: Async](scheduleTask: TriggeredV4 => F[Either[ErrorDetails, Boolean]],
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/consumers/TriggeredConsumer.scala]: Same points as above

File: src/main/scala/com/ovoenergy/orchestration/kafka/consumers/TriggeredConsumer.scala [hunk 11]
```diff
@@ -1,69 +1,100 @@
...
-      case Right(_) => Async[F].pure(())
-    }
+    def orchestrateOrFail(triggeredV4: TriggeredV4, internalMetadata: InternalMetadata): F[Unit] = {
+
+      def sendFeedbackIfFailure(either: Either[ErrorDetails, RecordMetadata],
+                                internalMetadata: InternalMetadata): F[Unit] = either match {
+        case Left(err) =>
+          warn(triggered)(s"Error orchestrating comm: ${err.reason}")
+          issueFeedback.sendWithLegacy(failureDetailsFromErr(err), triggeredV4.metadata, internalMetadata).void
+        case Right(_) => Async[F].pure(())
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/consumers/TriggeredConsumer.scala]: this is `F.unit` for any applicative, or `().pure[F]`

File: src/main/scala/com/ovoenergy/orchestration/kafka/consumers/TriggeredConsumer.scala [hunk 12]
```diff
@@ -1,69 +1,100 @@
...
-        warn(triggeredV4)(s"Error orchestrating comm: ${err.reason}")
-        sendFailedEvent(failedEventFromErr(err))
-      case Right(_) => Async[F].pure(())
-    }
+    def orchestrateOrFail(triggeredV4: TriggeredV4, internalMetadata: InternalMetadata): F[Unit] = {
+
+      def sendFeedbackIfFailure(either: Either[ErrorDetails, RecordMetadata],
+                                internalMetadata: InternalMetadata): F[Unit] = either match {
+        case Left(err) =>
+          warn(triggered)(s"Error orchestrating comm: ${err.reason}")
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/consumers/TriggeredConsumer.scala]: this log statement is not wrapped in `F`

File: src/main/scala/com/ovoenergy/orchestration/scheduling/TaskExecutor.scala [hunk 13]
```diff
@@ -44,7 +53,7 @@ object TaskExecutor extends LoggingWithMDC {
           case Left(err) => {
             warn(triggered)(s"Failed to orchestrate comm: ${err.reason}")
             IO(persistence.setScheduleAsFailed(scheduleId, err.reason))
-              .flatMap(_ => buildAndSendFailedEvent(err.reason, triggered, err.errorCode, internalMetadata))
+              .flatMap(_ => buildAndSendFailedEvents(triggered, err, internalMetadata))
               .map(_ => ())
           }
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/scheduling/TaskExecutor.scala]: `>>`, `.void`

File: src/main/scala/com/ovoenergy/orchestration/templates/RetrieveTemplateDetails.scala [hunk 14]
```diff
@@ -33,9 +33,9 @@ object RetrieveTemplateDetails extends LoggingWithMDC {
     def template(): ErrorsOr[CommTemplate[Id]] = {
       TemplatesRepo
         .getTemplate(templatesContext, templateManifest)
-        .leftMap{error =>
-            cachingStrategy.remove(templateId)
-            error
+        .leftMap { error =>
+          cachingStrategy.remove(templateId)
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/templates/RetrieveTemplateDetails.scala]: I don't know what's going on here, but this is side-effectful so longer term it should be changed

File: src/main/scala/com/ovoenergy/orchestration/kafka/consumers/CancellationRequestConsumer.scala [hunk 15]
```diff
@@ -15,50 +14,36 @@ import scala.concurrent.ExecutionContext
...
+              cancellationRequest,
+              s"Cancellation of scheduled comm failed: ${err.reason}"
             )
-            sendFailedCancellationEvent(
-              FailedCancellationV3(
-                GenericMetadataV3.fromSourceGenericMetadata("orchestration", cancellationRequest.metadata),
-                cancellationRequest,
-                s"Cancellation of scheduled comm failed: ${err.reason}"
-              )) >> issueFeedback.send(failureDetails)
+            F.pure(warn(cancellationRequest)(s"Cancellation request failed with error $err")) *>
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/consumers/CancellationRequestConsumer.scala]: ah, gave you some wrong advice here. Suspending things needs to be done with `F.delay` (`pure` doesn't suspend), and that requires a Sync bound. This problem will be fixed by a logging typeclass like the one in log4cats

File: src/main/scala/com/ovoenergy/orchestration/kafka/consumers/TriggeredConsumer.scala [hunk 16]
```diff
@@ -49,9 +49,9 @@ object TriggeredConsumer extends LoggingWithMDC {
       def sendFeedbackIfFailure(either: Either[ErrorDetails, RecordMetadata],
                                 internalMetadata: InternalMetadata): F[Unit] = either match {
         case Left(err) =>
-          warn(triggered)(s"Error orchestrating comm: ${err.reason}")
-          issueFeedback.sendWithLegacy(failureDetailsFromErr(err), triggeredV4.metadata, internalMetadata).void
-        case Right(_) => Async[F].pure(())
+          F.pure(warn(triggered)(s"Error orchestrating comm: ${err.reason}")) *>
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/orchestration/kafka/consumers/TriggeredConsumer.scala]: same here, as it turns out

**Comment by ovo-comms-circleci** [comment]
This PR was successfully deployed to production by [CircleCI build 522](https://circleci.com/gh/ovotech/comms-orchestration/522) at 2018-09-03 14:17 BST.

**Pull request merged**